File: python/src/behavioral/observer/concrete_observer_a.py
from .observer import Observer


class ConcreteObserverA(Observer):
    """
    Concrete Observer A reacts to subject updates.
    """

    def __init__(self):
        self.last_message: str | None = None

    def update(self, message: str) -> None:
        self.last_message = f"Observer A received: {message}"

File: go/src/behavioral/observer/concrete_observer_a.go
package observer

import "sync"

type ConcreteObserverA struct {
	mu       sync.RWMutex
	id       string
	lastSeen string
}

func NewConcreteObserverA(id string) *ConcreteObserverA {
	return &ConcreteObserverA{id: id}
}

func (o *ConcreteObserverA) ID() string {
	return o.id
}

func (o *ConcreteObserverA) Update(state string) {
	o.mu.Lock()
	o.lastSeen = "A received: " + state
	o.mu.Unlock()
}

func (o *ConcreteObserverA) LastSeen() string {
	o.mu.RLock()
	defer o.mu.RUnlock()
	return o.lastSeen
}

